### PR TITLE
feat: add assignment caching to android sdk

### PR DESCRIPTION
**Pull request by typotter**
Eppo Internal
:tickets: fixes FF-2280

## Motivation and Context
Deduplication of Assignment logs in the Android SDK

## Description of Changes
- new `LRUAssignmentCache` implementation of the `IAssignmentCache` interface
- `EppoClient.Init` uses LRU Assignment cache by default with a max size of 1024
- Also linted the example app

## How has this been tested
- Test cases in `EppoClientTest` to demonstrate duplicated logging calls and test deduplication.

## What is expected of reviewers
- 👀

### CODE REVIEW

File: eppo/src/main/java/cloud/eppo/android/EppoClient.java [hunk 1]
```diff
@@ -121,6 +125,9 @@ public static class Builder {
     private boolean offlineMode = false;
     private CompletableFuture<Configuration> initialConfiguration;
 
+    // Assignment caching on by default. To disable, call `builder.assignmentCache(null);`
+    private IAssignmentCache assignmentCache = new LRUAssignmentCache(10_000);
```
Comment (aarsilv) [inline eppo/src/main/java/cloud/eppo/android/EppoClient.java:129]: ❤️

File: eppo/src/main/java/cloud/eppo/android/EppoClient.java [hunk 2]
```diff
@@ -161,6 +168,11 @@ public Builder offlineMode(boolean offlineMode) {
       return this;
     }
 
+    public Builder assignmentCache(IAssignmentCache assignmentCache) {
```
Comment (aarsilv) [inline eppo/src/main/java/cloud/eppo/android/EppoClient.java:171]: 👍

File: eppo/src/main/java/cloud/eppo/android/cache/LRUAssignmentCache.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
+package cloud.eppo.android.cache;
+
+import android.util.LruCache;
```
Comment (aarsilv) [inline eppo/src/main/java/cloud/eppo/android/cache/LRUAssignmentCache.java:3]: 🔥

File: eppo/src/main/java/cloud/eppo/android/cache/LRUAssignmentCache.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+import cloud.eppo.api.IAssignmentCache;
+import cloud.eppo.logging.Assignment;
+
+public class LRUAssignmentCache implements IAssignmentCache {
```
Comment (aarsilv) [inline eppo/src/main/java/cloud/eppo/android/cache/LRUAssignmentCache.java:9]: Per the other PR, ideally this would extend the abstract cache

File: example/build.gradle [hunk 5]
```diff
@@ -47,3 +48,22 @@ dependencies {
...
 }
+
+spotless {
+    format 'misc', {
+        // define the files to apply `misc` to
+        target '*.gradle', '.gitattributes', '.gitignore'
+
+        // define the steps to apply to those files
+        trimTrailingWhitespace()
+        indentWithSpaces(2) // or spaces. Takes an integer argument if you don't like 4
```
Comment (aarsilv) [inline example/build.gradle:59]: this comment seems outdated

File: example/build.gradle [hunk 6]
```diff
@@ -1,5 +1,6 @@
 plugins {
     id 'com.android.application'
+    id "com.diffplug.spotless" version "7.0.0.BETA2"
```
Comment (aarsilv) [inline example/build.gradle:3]: 💪 
Like that you're adding a linter to enforce code style consistency but typically those things are done best as separate PRs so linting changes are not intermixed with actual changes

**Review by aarsilv — COMMENTED** [review]
Things are looking good! Approval dependent on changes with the downstream common SDK. Also recommend breaking linter changes out into their own PR.

File: eppo/src/androidTest/java/cloud/eppo/android/EppoClientTest.java [hunk 7]
```diff
@@ -80,6 +82,30 @@ private void initClient(
       @Nullable ConfigurationStore configurationStoreOverride,
       String apiKey,
       boolean offlineMode) {
+    initClient(
```
Comment (aarsilv) [inline eppo/src/androidTest/java/cloud/eppo/android/EppoClientTest.java:85]: How come we need this extra method in this test. Seems like we could just pass null (or a mock) as desired

File: eppo/src/androidTest/java/cloud/eppo/android/EppoClientTest.java [hunk 8]
```diff
@@ -651,6 +678,45 @@ public void testAssignmentEventCorrectlyCreated() {
...
+
+    Attributes subjectAttributes = new Attributes();
+    subjectAttributes.put("age", EppoValue.valueOf(30));
+    subjectAttributes.put("employer", EppoValue.valueOf("Eppo"));
+
+    EppoClient.getInstance().getDoubleAssignment("numeric_flag", "alice", subjectAttributes, 0.0);
+    EppoClient.getInstance().getDoubleAssignment("numeric_flag", "alice", subjectAttributes, 0.0);
+
+    ArgumentCaptor<Assignment> assignmentLogCaptor = ArgumentCaptor.forClass(Assignment.class);
+    verify(mockAssignmentLogger, times(1)).logAssignment(assignmentLogCaptor.capture());
```
Comment (aarsilv) [inline eppo/src/androidTest/java/cloud/eppo/android/EppoClientTest.java:717]: 👍 

I think the flip-flop case is covered in the common repo test 👌

File: example/build.gradle (hunk 5 above)
Comment (aarsilv) [reply example/build.gradle]: For Java I like 4 but it's not a strong preference nor am I am a JVM expert. I defer to @felipecsl's wisdom on indentation!

**Review by aarsilv — APPROVED** [review]
Thanks for iterating on the feedback! 💪 💪 💪 
image

File: eppo/src/androidTest/java/cloud/eppo/android/EppoClientTest.java (hunk 7 above)
Comment (typotter) [reply eppo/src/androidTest/java/cloud/eppo/android/EppoClientTest.java]: there were sooo many callsites and I was being lazy.

File: example/build.gradle (hunk 5 above)
Comment (typotter) [reply example/build.gradle]: this is actually for all the non .java files. We're applying the "Google java style" to the actual java files which is 2 spaces
